modules/nf-core/limma/differential/templates/limma_de.R

```diff
@@ -282,48 +300,64 @@ if ((! is.null(opt\$exclude_samples_col)) && (! is.null(opt\$exclude_samples_val
 ################################################
 ################################################
 
-# Build the model formula with blocking variables first
-model_vars <- c()
+if (!is.null(opt\$formula)) {
+    model <- opt\$formula
+} else {
+
+    # Build the model formula with blocking variables first
+    model_vars <- c()
 
-if (!is.null(opt\$blocking_variables)) {
-    # Include blocking variables (including pairing variables if any)
-    model_vars <- c(model_vars, blocking.vars)
-}
+    if (!is.null(opt\$blocking_variables)) {
+        # Include blocking variables (including pairing variables if any)
+        model_vars <- c(model_vars, blocking.vars)
+    }
 
-# Add the contrast variable at the end
-model_vars <- c(model_vars, contrast_variable)
+    # Add the contrast variable at the end
+    model_vars <- c(model_vars, contrast_variable)
 
-# Construct the model formula
-model <- paste('~ 0 +', paste(model_vars, collapse = '+'))
+    # Construct the model formula
+    model <- paste('~ 0 +', paste(model_vars, collapse = '+'))
 
-# Make sure all the appropriate variables are factors
-vars_to_factor <- model_vars  # All variables in the model need to be factors
-for (v in vars_to_factor) {
-    sample.sheet[[v]] <- as.factor(sample.sheet[[v]])
-}
+    # Make sure all the appropriate variables are factors
+    vars_to_factor <- model_vars  # All variables in the model need to be factors
+    for (v in vars_to_factor) {
+        sample.sheet[[v]] <- as.factor(sample.sheet[[v]])
+    }
 
+}
 ################################################
 ################################################
 ## Run Limma processes                        ##
 ################################################
 ################################################
+if (!is.null(opt\$formula)) {
```

Comment:
This is an if with exactly the same condition as the one above. I would merge them.